### ISSUE TITLE
How to write a custom loss function for a classifier

### ISSUE THREAD

**Issue by lesshaste** [EpistasisLab/tpot#1169]
I am using tpot  for binary classification.  I would like to optimize the true positive rate when the false positive rate is exactly zero.  Is there a way to define this objective function so that tpot can optimize for it?

**Comment by JDRomano2** [comment]
I think you might be confusing _loss functions_ with _scoring functions_: Loss functions are an internal attribute of some classification or regression operators. Scoring functions are applied to TPOT pipelines to assess the performance of that entire pipeline.

TPOT does not interact with loss functions (edit: unless you include them as a tunable metaparameter), so there is no way to customize loss functions in the context of TPOT. You can, in theory, define a custom classifier that uses a custom loss function, and then add that custom operator to TPOT via a custom configuration dictionary. But also keep in mind that many classification algorithms don't even use loss functions in their learning process. 

If, however, you want to use a custom _scoring_ function (instead of a loss function), details for how to do this are given in the documentation: https://epistasislab.github.io/tpot/using/#scoring-functions. You can use one of the predefined scoring metrics from scikit-learn (listed at https://scikit-learn.org/stable/modules/model_evaluation.html#common-cases-predefined-values), or you can define your own Python function that accepts a list of y values and predicted y values and returns a score.

**Comment by lesshaste** [comment]
Thank you. If I define a custom score function will tpot use that to optimize over pipelines?  Do score functions just have to output a number or does it have to be in some range?

**Comment by JDRomano2** [comment]
TPOT will use a custom score function if you do the following:

1). Define the function in another local module. (e.g., save it in `./score_f.py`)

2.) Give the function the signature `scorer(estimator, X, y)`

3.) Pass the import statement for the argument to the `scorer` argument of `TPOTClassifier()`. (e.g., `scorer='score_f.my_scorer`, if the function is named `my_scorer` and saved in `score_f.py`)

TPOT considers pipelines with higher scores to be better, so if your metric corresponds to "smaller is better" your scoring function for that metric should be negated.

Regarding your second question, it does not have to be in a specific range, but should be a floating point number.
